---
title: "MuRF1 Timecourse DGE report"
output: 
 html_notebook:
  toc: yes
---
  
# Clustering - Zea Mays Meoisis Riboseq data
  
# Setup
  
```{r loadinglibraries, include=FALSE, echo=FALSE, eval=T}
{
knitr::opts_chunk$set(root.dir = here::here(),eval=TRUE,cache=FALSE,echo=FALSE,warning = FALSE,message = FALSE,include=TRUE,
                      fig.width =7,fig.height=7,out.width=700,out.height=700,dev='png')

library(rmarkdown)
library(knitr)
library(here)
library(magrittr)
library(stringr)
library(ggplot2)
library(tximport)
library(dbscan)

dir.create(here('data'),showWarnings=F)
dir.create(here('R_cache'),showWarnings=F)

base::source(here('src/Rprofile.R'))

l2fc_threshold <- 0.32
}
```

```{r dds, include=FALSE, echo=FALSE, eval=T}
{

#this perl creates a file with unique gene_id, gene_name in pairs, faster than loading the whole gtf if it's big
# perl -lane '/gene_name\W+(\w+)\W/ ;my $a=$1;/gene_id\W+(\w+)\W/; print($a,"\t",$1)' <(cat  pipeline/genes.chrnm.gtf) | uniq | sort | uniq > pipeline/gid_2_gname.txt
# gtf <- here('pipeline/genes.chrnm.gtf')
# anno <- projmemoise(function(gtf){rtracklayer::import(gtf)})(gtf)
gid2gname <- read_tsv(here('data/gid_2_gname.txt'))%>%set_colnames(c('gene_name','gene_id'))
tx2gene <-  read_tsv(here('pipeline/gid_2_trid.txt.gz'))%>%set_colnames(c('g_id','tr_id'))

#gid2gname <- read_tsv(here('data/gid_2_gname.txt'))%>%set_colnames(c('gnm','g_id'))

#read in the sample_parameter file from the piepline
sampdf <- fread(here('src/sample_parameter.csv'))%>%
  select(-(library_layout:fragment_length_sd))%>%
  as.data.frame%>%set_rownames(.,.$sample_id)
#use the factor ordering in the sample_parameters file - by default it gets

sampdf <- sampdf%>%mutate_if(is_character,as_factor)
sampdf$assay%<>%fct_relevel(c('total','ribo'))
stopifnot(!any(is.na(sampdf$assay)))
#
if(is.null(sampdf$sample_name)) sampdf$sample_name <- sampdf$sample_id
#read the counts data - this takes a while sometimes
# file.remove(here('data/dds.rds'))


if(!file.exists(here('data/drimseqob.rds'))){
   
  iso_tx_countdata <- readRDS(here('data/iso_tx_countdata.rds'))
  #create our Drimseq objects
  cts <- iso_tx_countdata$counts
  cts <- cts[rowSums(cts) > 0,]

  txdf = tx2gene%>%select(GENEID=g_id,TXNAME=tr_id)%>%group_by(GENEID)%>%mutate(ntx=n())
#Run drimseq
  
  dds <- readRDS(here('data/dds.rds'))
  allcountdesign=colData(dds)%>%as.data.frame
  if(!'sample_id'%in%colnames(allcountdesign)) allcountdesign%<>%mutate(sample_id=sample)
  
  txdf <- txdf[match(rownames(cts),txdf$TXNAME),]
  all(rownames(cts) == txdf$TXNAME)


  {
  counts <- data.frame(gene_id=txdf$GENEID,
                       feature_id=txdf$TXNAME,
                       cts)
  library(DRIMSeq)
  drimseqob <- dmDSdata(counts=counts, samples=allcountdesign%>%as.data.frame)
  n.small = allcountdesign%>%group_by(time,assay)%>%tally%>%.$n%>%min
  ncountrows = nrow(allcountdesign)
  drimseqob <- dmFilter(drimseqob,
                min_samps_feature_expr=n.small, min_feature_expr=10,
                min_samps_feature_prop=n.small, min_feature_prop=0.1,
                min_samps_gene_expr=ncountrows, min_gene_expr=10)
  design_full <- model.matrix(~group, data=DRIMSeq::samples(drimseqob))
 
  sgroupvects = model.matrix(data=dds@colData,design(dds))%>%set_rownames(dds@colData$group)%>%{.[unique(rownames(.)),]}%>%
    {m=.;lapply(1:nrow(.)%>%setNames(rownames(m)),function(i){m[i,]}) }

contrasts <- list(
  "MT_vs_MB" = with(sgroupvects,MT - MB),
  "Ctrl_MT_A24_vs_MT" = with(sgroupvects,MT_Ctrl24 - MT),
  "Ctrl_MT_A72_vs_MT_A24" = with(sgroupvects,MT_Ctrl72 - MT_Ctrl24),
  "Ctrl_MT_A72_vs_MT0" = with(sgroupvects,MT_Ctrl72 - MT),
  "Aspecific_MT_A24_vs_MT" = with(sgroupvects,(MT_A24 - MT) - (MT_Ctrl24 - MT)),
  "Aspecific_MT_A72_vs_MT_A24" = with(sgroupvects,(MT_A72 - MT_A24) - (MT_Ctrl72 - MT_Ctrl24)),
  "Aspecific_MT_A72_vs_MT0" = with(sgroupvects,(MT_A72 - MT) - (MT_Ctrl72 - MT))
)

  drimseqob = drimseqob
  drimseqob <- dmPrecision(drimseqob, design=design_full)
  drimseqob <- dmFit(drimseqob, design=design_full)
  drimseqob <- dmTest(drimseqob, contrasts[[1]])
  drimseqob2 <- dmTest(drimseqob, contrasts[[1]])

  } 
  saveRDS(drimseqob,here('data/drimseqob.rds'))
}else{
  drimseqob<-readRDS(here('data/drimseqob.rds'))
}

cat('successfully loaded  countdata and annotation')
stopifnot(exists('sampdf'))
}
```


```{r get normalized counts}
#you might want to use vst if there's a lot of data, for speed
dir.create(here('data'))
if(!file.exists(here('data/normcounts.rds'))){
  
  normfunc <- if(ncol(dds)>20) DESeq2::rlog else DESeq2::vst
  
  normcounts <- normfunc(dds)
  rownames(normcounts) <- rownames(dds)
  saveRDS(normcounts,here('data/normcounts.rds'))
}else{
  normcounts<-readRDS(here('data/normcounts.rds'))
}
cat('successfully normalized  countdata')
```


```{r test contrasts}
stopifnot(exists('dds'))
stopifnot(exists('normcounts'))
stopifnot(exists('sampdf'))
{
  #do a trial run of DESeq to get the contrast names
  design <- as.formula('~ group')
  design(dds) = design
  testdds = DESeq(head(dds,2*ncol(dds)))
  resnames = resultsNames(testdds)
  
}
# [1] "Intercept"             "group_MT_vs_MB"        "group_MT_A24_vs_MB"
# [4] "group_MT_A72_vs_MB"    "group_MT_Ctrl24_vs_MB" "group_MT_Ctrl72_vs_MB"
resnames = resultsNames(testdds)
stopifnot(resnames==c("Intercept", "group_MT_vs_MB", "group_MT_A24_vs_MB", "group_MT_A72_vs_MB", 
"group_MT_Ctrl24_vs_MB", "group_MT_Ctrl72_vs_MB"))

sgroupvects = model.matrix(data=dds@colData,design(dds))%>%set_rownames(dds@colData$group)%>%{.[unique(rownames(.)),]}%>%
{m=.;lapply(1:nrow(.)%>%setNames(rownames(m)),function(i){m[i,]}) }

contrasts <- list(
  "MT_vs_MB" = with(sgroupvects,MT - MB),
  "Ctrl_MT_A24_vs_MT" = with(sgroupvects,MT_Ctrl24 - MT),
  "Ctrl_MT_A72_vs_MT_A24" = with(sgroupvects,MT_Ctrl72 - MT_Ctrl24),
  "Aspecific_MT_A24_vs_MT" = with(sgroupvects,(MT_A24 - MT) - (MT_Ctrl24 - MT)),
  "Aspecific_MT_A72_vs_MT_A24" = with(sgroupvects,(MT_A72 - MT_A24) - (MT_Ctrl72 - MT_Ctrl24))
)

#check the contrasts
for(contrastname in names(contrasts)){
  contrast = contrasts[[contrastname]]
  if(is.character(contrast) & length(contrast)==1){
    assert_that(contrast %in% resnames)
    contrast <- as.numeric(resnames==contrast)
  }
  message(contrastname)
  try({
    results(testdds,contrast)
  })
}

cat('tested contrasts')
```




# Individual gene TR use plots

```{r runcontrasts}
gid2gname <- read_tsv(here('data/gid_2_gname.txt'))%>%set_colnames(c('gnm','g_id'))
tx2gene <-  read_tsv(here('pipeline/gid_2_trid.txt.gz'))%>%set_colnames(c('g_id','tr_id'))

igene_id = gid2gname%>%filter(gnm=='Trim63')%>%.$g_id
plotProportions(drimseqob, gene_id = igene_id, group_variable = "group",
plot_type = "ribbonplot")



  
igene = 'Trim63'
igeneid = gid2gname %>% 
  safe_filter(gnm==igene)%>%
  .$g_id
igenetrs = tx2gene %>% 
  safe_filter(g_id==igeneid)%>%
  .$tr_id%>%unique
dsets = iso_tx_countdata$abundance%>%colnames
colgrps = dsets%>%str_extract('.*?(?=_\\d+)')
colgrp = colgrps[[1]]

ucolgrps = unique(colgrps)%>%setNames(.,.)
relativeusedf = map_df(.id='dset',ucolgrps,function(colgrp){
  sab = iso_tx_countdata$abundance[,colgrps==colgrp]%>%rowSums
  sab%>%
    enframe('tr_id','abundance')%>%
    left_join(tx2gene%>%select(g_id,tr_id),by='tr_id')%>%
    left_join(gid2gname,by='g_id')%>%
    group_by(gnm)%>%mutate(abundance=abundance/sum(abundance))
})

ign_usedf = relativeusedf%>%
  safe_filter(gnm==igene)%>%
  # separate(dset,c('time','assay'))
  identity

#now plot
plotfile<- here(paste0('plots/',,'.pdf'))
pdf(plotfile)
locusplot(igene)
dev.off()
message(normalizePath(plotfile))


```







```{r runcontrasts}
#you might want to use vst if there's a lot of data, for speed
dir.create(here('data'))
if(!file.exists(here('data/resultslist.rds'))){
	design(dds) <- design
 
  dds <- DESeq(dds,betaPrior = F)
  #and get the contrasts we need
  resultslist <- lapply(contrasts,results,object = dds)
  names(resultslist) <- names(contrasts)
  for(i in seq_along(resultslist))resultslist[[i]]$gene_id = rownames(dds)
  
  ddstodf<- function(ddsdf)results(ddsLRT)%>%as.data.frame%>%rownames_to_column('gene_id')
  ddsLRT = DESeq(dds, test="LRT", full=~group,reduced = ~ 1)
  changegenes <- results(ddsLRT)%>%ddstodf %>%subset%>%filter(padj < 0.05)%>%.$gene_id
  saveRDS(changegenes,here('data/changegenes.rds'))
  
  saveRDS(resultslist,here('data/resultslist.rds'))
  
  
  
}else{
  resultslist<-readRDS(here('data/resultslist.rds'))
  changegenes<-readRDS(here('data/changegenes.rds'))
}

cat('successfully calculated contrasts:')
message(names(resultslist)%>%paste(collapse=' '))

resultsall = resultslist%>%map_df(.id='contrast',as.data.frame)%>%left_join(gid2gname)
resultsall%>%group_by(contrast)%>%filter(padj<0.05,abs(log2FoldChange)>log2(1.5))%>%mutate(direction=ifelse(log2FoldChange>0,'up','down'))%>%group_by(contrast,direction)%>%tally%>%kable(main='differentially Expressed Genes')

```


```{r}
#get highcount genes
highcountgenes<-counts(dds)%>%as.data.frame%>%rownames_to_column('gene_id')%>%gather(lib,count,-gene_id)%>%
    mutate(lib = str_replace(lib,'_\\d$',''))%>%
  {stopifnot(tally(.)%>%.$n%>%table%>%length%>%`==`(1));.}%>%
  group_by(gene_id,lib)%>%
    summarise(count=sum(count))%>%
    filter(count>32)%>%
  .$gene_id

changecols<-c('#1db220','#737373','#fef636','#b31261')%>%setNames(c("TE only", "mRNA only", "compensating", "concurrent"))

# #data set for fold changes
# fcsetdf = resultslist$%>%select(TEpadj =padj, TEFC = log2FoldChange,gene_id)%>%
#     left_join(resultslist$fraction_si_vs_mock%>%as.data.frame%>%
#                 select(TRpadj =padj, TRFC = log2FoldChange,gene_id))%>%
#     filter(gene_id %in% highcountgenes)%>%
#     mutate(
#       TEchange = (sign(TEFC))*as.numeric(TEpadj<0.05),
#       TRchange = (sign(TRFC))*as.numeric(TRpadj<0.05),
#       set = case_when(
#        ((TEchange * TRchange) == -1)  ~ 'compensating',
#       (TEchange * TRchange) == 1  ~ 'concurrent' ,
#         (abs(TRchange)) == 1 ~ "mRNA only",
#         (abs(TEchange)) == 1 ~ "TE only",
#       TRUE ~ as.character(NA)
#       )
#     )

#print fc plot
# dir.create(here('tables'),showWarnings = F)
# fcsetdf%>%left_join(gid2gname%>%select(gene_id,gene_name))%>%write_tsv(here('tables/fcsetdf.tsv')%T>%normalizePath%T>%message)
# p = fcsetdf %>%
#   filter(!is.na(set))%>%
#   qplot(data=.,x=TRFC,y=TEFC,geom='point',color=set)+
#   theme_bw()+
#   scale_color_manual('Set',values = changecols)+
#   scale_x_continuous('Transcriptional change',limits=c(-3,3))+
#   scale_y_continuous('TE change (xtail)',limits=c(-3,3))+
#   coord_fixed()
# plotfile <- 'plots/xtail_fc.pdf'
# dir.create(dirname(plotfile))
# pdf(plotfile)
# print(p)
# dev.off()
# message(normalizePath(plotfile))

```


## Contrasts Enrichment

```{r,eval=F}




base::source(here('src/gofuncs.R'))
GTOGO <- select(GTOGO,gene_name,go_id,gene_id)%>%filter(gene_id %in% (rownames(normcounts)%>%str_replace(.,'\\.\\d+','')))%>%group_by(go_id)%>%filter(n()>10)

regdirfuncs = list(Up=identity,Down=function(x)x * -1)

fc_df <- resultslist[[1]]
regdirfunc=regdirfuncs[[1]]
ont='BP'
if(!file.exists(here('data/contrgotables.rds'))){
  contrgotables<-map_df(.id='contrast',resultslist,function(fc_df){
    map_df(.id='regdir',regdirfuncs,function(regdirfunc){
      map_df(.id='ontology',c('BP','MF','CC')%>%setNames(.,.),function(ont){
         possibly(rungo,otherwise = NULL,quiet=TRUE)(fc_df%>%{setNames((regdirfunc(.$log2FoldChange)>0) & (.$padj<0.05),.$gene_id)},GTOGO,ont)
      })
    })
  })
  saveRDS(contrgotables,here('data/contrgotables.rds'))
}else{
  contrgotables<-readRDS(here('data/contrgotables.rds'))
}

```


```{r,eval=}

for(contr in unique(contrgotables$contrast)){
  for(regdir in unique(contrgotables$regdir)){
    for(ont in unique(contrgotables$ontology)){
        plot_go_enrich(contrgotables%>%filter(contrast==contr,regdir==regd,ontology==ont),'elimFisher',paste(contrast,regdir,ontology))
    }
  }
}

```


```{r,eval=F}
library(DT)
datatable(contrgotables%>%select(ontology,contrast,regdir,GO.ID,elimFisher,Term,Enrichment,Annotated,Significant,Expected),filter=c('top'))
```


# Output Tables

Below you can see the files into which various objects from the above analysis have been saved.

```{r, echo=TRUE,include=TRUE}
dir.create(here('tables'),showWarnings = F)

#raw count data
dir.create('tables')
counts(dds)%>%as.data.frame%>%rownames_to_column('gene_id')%>%write_tsv(here('tables/rawcountdata.tsv'))
tx_countdata$abundance%>%as.data.frame%>%rownames_to_column('gene_id')%>%left_join(gid2gname)%>%write_tsv(here('tables/tpms.tsv'))
#all coefficient, fold changes
resultslist%>%map_df(.id='contrast',as.data.frame)%>%left_join(gid2gname,by='gene_id')%>%write_tsv(here('tables/allcontrasts.tsv'))

# #mapping from GO terms to genes
# GTOGO%>%write_tsv(here('tables/go_gene_map.tsv'))
# #
# contrgotables%>%write_tsv(here('tables/contrast_goenrich.tsv'))

