---
title: "MuRF1 Timecourse DGE report"
output: 
 html_notebook:
  toc: yes
---
  
# Clustering - Zea Mays Meoisis Riboseq data
  
# Setup
  
```{r loadinglibraries, include=FALSE, echo=FALSE, eval=T}
knitr::opts_chunk$set(root.dir = here::here(),eval=TRUE,cache=FALSE,echo=FALSE,warning = FALSE,message = FALSE,include=TRUE,
                      fig.width =7,fig.height=7,out.width=700,out.height=700,dev='png')

library(rmarkdown)
library(knitr)
library(here)
library(magrittr)
library(stringr)
library(ggplot2)
library(dbscan)

dir.create(here('data'),showWarnings=F)
dir.create(here('R_cache'),showWarnings=F)

base::source(here('src/Rprofile.R'))


```


```{r}
l2fc_threshold <- 0.32
```

```{r dds, include=FALSE, echo=FALSE, eval=T}

#this perl creates a file with unique gene_id, gene_name in pairs, faster than loading the whole gtf if it's big
# perl -lane '/gene_name\W+(\w+)\W/ ;my $a=$1;/gene_id\W+(\w+)\W/; print($a,"\t",$1)' <(cat  pipeline/genes.chrnm.gtf) | uniq | sort | uniq > pipeline/gid_2_gname.txt
# gtf <- here('pipeline/genes.chrnm.gtf')
# anno <- projmemoise(function(gtf){rtracklayer::import(gtf)})(gtf)
gid2gname <- fread(here('pipeline/gid_2_gname.txt'))%>%set_colnames(c('gene_name','gene_id'))
tx2gene <-  read_tsv(here('pipeline/trid2gid.txt'))%>%set_colnames(c('tr_id','g_id'))

#read in the sample_parameter file from the piepline
sampdf <- fread(here('src/sample_parameter.csv'))%>%
  select(-(library_layout:fragment_length_sd))%>%
  as.data.frame%>%set_rownames(.,.$sample_id)
#use the factor ordering in the sample_parameters file - by default it gets

sampdf <- sampdf%>%mutate_if(is_character,as_factor)
sampdf$assay%<>%fct_relevel(c('total','ribo'))
stopifnot(!any(is.na(sampdf$assay)))
#
if(is.null(sampdf$sample_name)) sampdf$sample_name <- sampdf$sample_id
#read the counts data - this takes a while sometimes
# file.remove(here('data/dds.rds'))


if(!file.exists(here('data/dds.rds'))){
   


  allquantfiles = Sys.glob(here::here('pipeline/salmon/data/*/quant.sf'))
  names(allquantfiles) <- allquantfiles%>%dirname%>%basename

  tx_countdata = tximport(files=allquantfiles,
    ignoreTxVersion=TRUE,
    tx2gene=tx2gene,
    type='salmon'
  )

  iso_tx_countdata = tximport(files=allquantfiles,
    ignoreTxVersion=TRUE,
    txOut=TRUE,
    tx2gene=tx2gene,
    type='salmon',
    importer =   function(file){read_tsv(file,col_types=cols())%>%
        mutate(Name=str_extract(Name,'ENSG\\w+|ENSMUST\\w+'))
      }
  )

  dds <- DESeqDataSetFromTximport(tx_countdata, 
    # sampdf%>%mutate(group=paste0(cell_type,'_',time)%>%str_replace('_0','')),
    sampdf%>%mutate(group=group%>%str_replace('_\\d$','')),
    ~ group )
  # colnames(counts(dds))<-sampdf$sample_id
  saveRDS(dds,here('data/dds.rds'))
}else{
  dds <-readRDS(here('data/dds.rds'))
}
cat('successfully loaded  countdata and annotation')
stopifnot(exists('sampdf'))
```


```{r get normalized counts}
#you might want to use vst if there's a lot of data, for speed
dir.create(here('data'))
if(!file.exists(here('data/normcounts.rds'))){
  
  normfunc <- if(ncol(dds)>20) DESeq2::rlog else DESeq2::vst
  
  normcounts <- normfunc(dds)
  rownames(normcounts) <- rownames(dds)
  saveRDS(normcounts,here('data/normcounts.rds'))
}else{
  normcounts<-readRDS(here('data/normcounts.rds'))
}
cat('successfully normalized  countdata')
```


```{r test contrasts}
stopifnot(exists('dds'))
stopifnot(exists('normcounts'))
stopifnot(exists('sampdf'))
{
  #do a trial run of DESeq to get the contrast names
  design <- as.formula('~ group')
  design(dds) = design
  testdds = DESeq(head(dds,2*ncol(dds)))
  resnames = resultsNames(testdds)
  
}
# [1] "Intercept"             "group_MT_vs_MB"        "group_MT_A24_vs_MB"
# [4] "group_MT_A72_vs_MB"    "group_MT_Ctrl24_vs_MB" "group_MT_Ctrl72_vs_MB"
resnames = resultsNames(testdds)

sgroupvects = model.matrix(data=dds@colData,design(dds))%>%set_rownames(dds@colData$group)%>%{.[unique(rownames(.)),]}%>%
{m=.;lapply(1:nrow(.)%>%setNames(rownames(m)),function(i){m[i,]}) }

contrasts <- list(
  "MT_vs_MB" = with(sgroupvects,MT - MB),
  "Ctrl_MT_A24_vs_MT" = with(sgroupvects,MT_Ctrl24 - MT),
  "Ctrl_MT_A72_vs_MT_A24" = with(sgroupvects,MT_Ctrl72 - MT_Ctrl24),
  "Aspecific_MT_A24_vs_MT" = with(sgroupvects,(MT_A24 - MT) - (MT_Ctrl24 - MT)),
  "Aspecific_MT_A72_vs_MT_A24" = with(sgroupvects,(MT_A72 - MT_A24) - (MT_Ctrl72 - MT_Ctrl24))
)

#check the contrasts
for(contrastname in names(contrasts)){
  contrast = contrasts[[contrastname]]
  if(is.character(contrast) & length(contrast)==1){
    assert_that(contrast %in% resnames)
    contrast <- as.numeric(resnames==contrast)
  }
  message(contrastname)
  try({
    results(testdds,contrast)
  })
}

cat('tested contrasts')
```


```{r runcontrasts}
#you might want to use vst if there's a lot of data, for speed
dir.create(here('data'))
if(!file.exists(here('data/resultslist.rds'))){
	design(dds) <- design
 
  dds <- DESeq(dds,betaPrior = F)
  #and get the contrasts we need
  resultslist <- lapply(contrasts,results,object = dds)
  names(resultslist) <- names(contrasts)
  for(i in seq_along(resultslist))resultslist[[i]]$gene_id = rownames(dds)
  
  ddstodf<- function(ddsdf)results(ddsLRT)%>%as.data.frame%>%rownames_to_column('gene_id')
  ddsLRT = DESeq(dds, test="LRT", full=~group,reduced = ~ 1)
  changegenes <- results(ddsLRT)%>%ddstodf %>%subset%>%filter(padj < 0.05)%>%.$gene_id
  saveRDS(changegenes,here('data/changegenes.rds'))
  
  saveRDS(resultslist,here('data/resultslist.rds'))
  
  
  
}else{
  resultslist<-readRDS(here('data/resultslist.rds'))
  changegenes<-readRDS(here('data/changegenes.rds'))
}

cat('successfully calculated contrasts:')
message(names(resultslist)%>%paste(collapse=' '))


```


# QC {.tabset}
```{r qc, include = TRUE,message=TRUE,eval=T,cache=F,results='asis'}
stopifnot(exists('dds'))
stopifnot(exists('normcounts'))
stopifnot(exists('sampdf'))

cat("## dispplot")
#plotDispEsts(dds)
rld <- normcounts
cat("## outlierplot")
if(is.null(SummarizedExperiment::assays(dds)[["cooks"]])){
  nulltext = "Cooks distance is meaningless\n because we only have one group per sample"
  # qplot(label=nulltext,x=1,geom='text',y=1,size=I(14))+theme_bw()
}else{
  boxplot(log10(SummarizedExperiment::assays(dds)[["cooks"]]), range=0, las=2)
}

cat("## countdist ,warning=F")
my_counts <-
  DESeq2::counts(dds)%>%
  as.data.frame() %>%
  tibble::rownames_to_column("feature_id") %>%
  tidyr::gather(sample_id, count, -feature_id)

sample_annot <- as.data.frame(SummarizedExperiment::colData(dds))
if(!'sample_name' %in% colnames(sample_annot)) sample_annot$sample_name<-sample_annot$sample_id
my_counts <-
  dplyr::left_join(my_counts, sample_annot, by = "sample_id")

stopifnot(exists('dds'))

my_counts %>%
  ggplot(aes(x = count + 1e-3, color = group, group = sample_name)) +
  geom_density() +
  scale_x_log10()

my_counts %>%
  dplyr::arrange(group) %>%
  dplyr::mutate(sample_name = factor(sample_name, levels = unique(sample_name))) %>%
  ggplot(aes(y = count + 1e-3, color = group, x = sample_name)) +
  geom_violin() +
  scale_y_log10() +
  coord_flip()


cat("## rlddf")
stopifnot(is(rld,'DESeqTransform'))

rld_df <- assay(rld)

rld_df <-
  assay(rld) %>%
  as.data.frame() %>%
  tibble::rownames_to_column('feature_id') %>%
  tibble::as_data_frame() %>%
  tidyr::gather(sample_id, reg_log_count, -feature_id)

rld_df$sample_name <- sample_annot$sample_name[match(rld_df$sample_id,colData(dds)$sample_id)]
# rld_df <- get_reg_log_counts(dds, blind = T)

rld_dispersion <-
  rld_df  %>%
  dplyr::group_by(feature_id) %>%
  dplyr::summarise(mean_reg_log_count = mean(reg_log_count),
                   sd_reg_log_count = sd(reg_log_count))

rld_dispersion %>%
  ggplot(aes(y = sd_reg_log_count, x = mean_reg_log_count)) +
  geom_point() +
  labs(x = "mean log regularized counts", y = "standar deviation log regularized counts")


cat("## pca, cache = F, fig_width=12,fig_height=12")
stopifnot(colnames(rld)==colnames(dds))

message('calculating PCAs')
pca = DESeq2::plotPCA(rld,intgroup="group")
pca = pca +
  expand_limits(x = pca$data$PC1 %>%
                  range %>%
                  multiply_by(1.5)) +
  expand_limits(y = pca$data$PC2 %>%
                  range %>%
                  multiply_by(1.3))
#(pca+geom_text(aes(group),size=I(2),alpha=I(0.5))) %>% {.$layers=.$layers[-1];.+geom_point(size=I(0))} +
#  ggtitle('PCA : labelled by samplename')
#we also want a table with these data to pick out outliers
pca$data %>% write_tsv(here('data/pca.tsv'))

cat("## pca_grouplabel, fig_width=12,fig_height=12")
(pca+geom_text(aes(label=group),size=I(2),alpha=I(0.5)))%>%{.$layers=.$layers[-1];.+geom_point(size=I(0))}+ggtitle('PCA : labelled by group')

cat("## pca_nolabel, fig_width=12,fig_height=12")
(pca+geom_text(aes(label=group),size=I(0),alpha=I(0.5)))%>%{.$layers=.$layers[-1];.+geom_point(size=I(3))}+ggtitle('PCA')


cat("## heatmap")
# dual scale settings for heatmaps
## setting values outside of the range [z_min, z_max] to the limiting values
if(!exists('z_max')) z_max <-  3.5
if(!exists('z_min')) z_min <- -3.5
# colnames(rld_df) %<>% {stringr::str_replace(.,'sample_id','sample_name')}
plot_heatmap_fluc_features(500, rld_df, z_min = z_min, z_max = z_max)

my_counts%<>%group_by(sample_name)%>%
  mutate(countclass = case_when(
    count > 1000 ~ '> 1000',
    count > 100 ~ '> 100',
    count > 32 ~ '> 32',
    count > 8 ~ '> 8',
    count > 0 ~ '< 8',
    count ==0 ~ ' 0 ',
    TRUE ~ 'NA'
  ))

my_counts$countclass%<>%factor(levels=c('> 1000','> 100','> 32','> 8','< 8',' 0 '))
stopifnot(!any(is.na(my_counts$countclass)))


my_counts%>%
  group_by(sample_name,countclass)%>%tally%>%spread(countclass,n)%>%.[rev(colnames(.))]%>%select(sample_name,everything())%>%kable(label='gene count classes')

```

qc plots above





# Differential expression

```{r, results='asis'}

trim_gids <- function(df){
    df=as.data.frame(df);
    df%>%mutate_at(vars(matches('gene_id|feature_id')),list(~str_replace(.,'\\.\\d+','')))
}
resultslist %<>% map(.%>%as.data.frame%>%trim_gids)

contrast = names(resultslist)[[4]]
regdirfuncs = list(Up=identity,Down=function(x)x * -1)
regdir='Up'

for(contrast in names(resultslist)){
  for(regdir in names(regdirfuncs)){
    regdirfunc = regdirfuncs[[regdir]]
    res = resultslist[[contrast]]%>%as.data.frame%>%trim_gids
     
    print(qplot(data = res, log10(baseMean), log2FoldChange , size=I(0.2), color= padj < 0.05)+theme_bw()+scale_color_manual(values=c('black','red')))
    
    
    library(DT)
      reg_features = res%>%filter(padj<0.05,regdirfunc(log2FoldChange) > l2fc_threshold) 
      num_reg <- nrow(reg_features)
      
      cat("${regdir} regulated feature: ${num_reg}")

      reg_features %>% select(-matches('contrast')) %>%
        left_join(gid2gname%>%select(gene_id,gene_name))%>%
        select(gene_name,gene_id,baseMean,log2FoldChange,padj,everything())%>%
        DT::datatable(rownames = F,escape = F,extensions = 'Buttons', 
                  options = list(
                    dom='Bfrtip',
                    buttons = list(
                      list(
                        extend='csv',
                        buttons=c('csv'),
                        text='download')
                    )
                  )
        )
  }
}
```


```{r}
#get highcount genes
highcountgenes<-counts(dds)%>%as.data.frame%>%rownames_to_column('gene_id')%>%gather(lib,count,-gene_id)%>%
    mutate(lib = str_replace(lib,'_\\d$',''))%>%
  {stopifnot(tally(.)%>%.$n%>%table%>%length%>%`==`(1));.}%>%
  group_by(gene_id,lib)%>%
    summarise(count=sum(count))%>%
    filter(count>32)%>%
  .$gene_id

changecols<-c('#1db220','#737373','#fef636','#b31261')%>%setNames(c("TE only", "mRNA only", "compensating", "concurrent"))

# #data set for fold changes
# fcsetdf = resultslist$%>%select(TEpadj =padj, TEFC = log2FoldChange,gene_id)%>%
#     left_join(resultslist$fraction_si_vs_mock%>%as.data.frame%>%
#                 select(TRpadj =padj, TRFC = log2FoldChange,gene_id))%>%
#     filter(gene_id %in% highcountgenes)%>%
#     mutate(
#       TEchange = (sign(TEFC))*as.numeric(TEpadj<0.05),
#       TRchange = (sign(TRFC))*as.numeric(TRpadj<0.05),
#       set = case_when(
#        ((TEchange * TRchange) == -1)  ~ 'compensating',
#       (TEchange * TRchange) == 1  ~ 'concurrent' ,
#         (abs(TRchange)) == 1 ~ "mRNA only",
#         (abs(TEchange)) == 1 ~ "TE only",
#       TRUE ~ as.character(NA)
#       )
#     )

#print fc plot
# dir.create(here('tables'),showWarnings = F)
# fcsetdf%>%left_join(gid2gname%>%select(gene_id,gene_name))%>%write_tsv(here('tables/fcsetdf.tsv')%T>%normalizePath%T>%message)
# p = fcsetdf %>%
#   filter(!is.na(set))%>%
#   qplot(data=.,x=TRFC,y=TEFC,geom='point',color=set)+
#   theme_bw()+
#   scale_color_manual('Set',values = changecols)+
#   scale_x_continuous('Transcriptional change',limits=c(-3,3))+
#   scale_y_continuous('TE change (xtail)',limits=c(-3,3))+
#   coord_fixed()
# plotfile <- 'plots/xtail_fc.pdf'
# dir.create(dirname(plotfile))
# pdf(plotfile)
# print(p)
# dev.off()
# message(normalizePath(plotfile))

```


## Contrasts Enrichment

```{r,eval=F}
base::source(here('src/gofuncs.R'))
GTOGO <- select(GTOGO,gene_name,go_id,gene_id)%>%filter(gene_id %in% (rownames(normcounts)%>%str_replace(.,'\\.\\d+','')))%>%group_by(go_id)%>%filter(n()>10)

regdirfuncs = list(Up=identity,Down=function(x)x * -1)

fc_df <- resultslist[[1]]
regdirfunc=regdirfuncs[[1]]
ont='BP'
if(!file.exists(here('data/contrgotables.rds'))){
  contrgotables<-map_df(.id='contrast',resultslist,function(fc_df){
    map_df(.id='regdir',regdirfuncs,function(regdirfunc){
      map_df(.id='ontology',c('BP','MF','CC')%>%setNames(.,.),function(ont){
         possibly(rungo,otherwise = NULL,quiet=TRUE)(fc_df%>%{setNames((regdirfunc(.$log2FoldChange)>0) & (.$padj<0.05),.$gene_id)},GTOGO,ont)
      })
    })
  })
  saveRDS(contrgotables,here('data/contrgotables.rds'))
}else{
  contrgotables<-readRDS(here('data/contrgotables.rds'))
}



```


```{r,eval=F}
for(contr in unique(contrgotables$contrast)){
  for(regdir in unique(contrgotables$regdir)){
    for(ont in unique(contrgotables$ontology)){
        plot_go_enrich(contrgotables%>%filter(contrast==contr,regdir==regd,ontology==ont),'elimFisher',paste(contrast,regdir,ontology))
    }
  }
}
```


```{r,eval=F}
library(DT)
datatable(contrgotables%>%select(ontology,contrast,regdir,GO.ID,elimFisher,Term,Enrichment,Annotated,Significant,Expected),filter=c('top'))
```


# Output Tables

Below you can see the files into which various objects from the above analysis have been saved.

```{r, echo=TRUE,include=TRUE}
dir.create(here('tables'),showWarnings = F)

#raw count data
counts(dds)%>%as.data.frame%>%rownames_to_column('gene_id')%>%write_tsv(here('tables/rawcountdata.tsv'))
assay(normcounts)%>%as.data.frame%>%rownames_to_column('gene_id')%>%write_tsv(here('tables/normcountdata.tsv'))
#all coefficient, fold changes
resultslist%>%map_df(.id='contrast',as.data.frame)%>%write_tsv(here('tables/allcontrasts.tsv'))

# #mapping from GO terms to genes
# GTOGO%>%write_tsv(here('tables/go_gene_map.tsv'))
# #
# contrgotables%>%write_tsv(here('tables/contrast_goenrich.tsv'))

